File: src/job.test.ts
import { readFile } from "node:fs/promises";

import { describe, expect, it } from "vitest";

import {
  CANONICAL_JOB_KINDS,
  JOB_FAILURE_REASONS,
  JOB_KIND_MESH_OPTIMIZATION,
  parseJobFailureReason,
  parseJobKind,
} from "./job.ts";

describe("job catalog", () => {
  it("matches the cross-language fixture", async () => {
    const fixture = JSON.parse(
      await readFile(
        new URL("../fixtures/job-catalog.json", import.meta.url),
        "utf8",
      ),
    ) as { job_kind: string; failure_reasons: string[] }[];

    expect(fixture).toEqual(
      CANONICAL_JOB_KINDS.map((jobKind) => ({
        job_kind: jobKind,
        failure_reasons: JOB_FAILURE_REASONS[jobKind],
      })),
    );
    expect(fixture.map(({ job_kind }) => parseJobKind(job_kind))).toEqual(
      CANONICAL_JOB_KINDS,
    );
    for (const { job_kind, failure_reasons } of fixture) {
      for (const reason of failure_reasons) {
        expect(parseJobFailureReason(parseJobKind(job_kind), reason)).toBe(
          reason,
        );
      }
    }
  });

  it("rejects unregistered values", () => {
    expect(() => parseJobKind("projection")).toThrow(
      "unknown canonical job kind",
    );
    expect(() =>
      parseJobFailureReason(JOB_KIND_MESH_OPTIMIZATION, "failed"),
    ).toThrow("unknown failure reason");
    expect(() =>
      parseJobFailureReason("projection" as never, "failed"),
    ).toThrow("unknown canonical job kind");
  });
});

File: job_test.go
package telemetry

import (
	"encoding/json"
	"os"
	"testing"
)

func TestJobCatalogMatchesCrossLanguageFixture(t *testing.T) {
	t.Parallel()
	data, err := os.ReadFile("fixtures/job-catalog.json")
	if err != nil {
		t.Fatal(err)
	}
	var fixture []struct {
		JobKind        string   `json:"job_kind"`
		FailureReasons []string `json:"failure_reasons"`
	}
	if err := json.Unmarshal(data, &fixture); err != nil {
		t.Fatal(err)
	}
	if len(fixture) != len(canonicalJobFailureReasons) {
		t.Fatalf("fixture has %d of %d job kinds", len(fixture), len(canonicalJobFailureReasons))
	}
	for _, entry := range fixture {
		jobKind, parseErr := ParseJobKind(entry.JobKind)
		if parseErr != nil {
			t.Fatal(parseErr)
		}
		if len(entry.FailureReasons) != len(canonicalJobFailureReasons[jobKind]) {
			t.Fatalf("fixture job %q has incomplete failure reasons", entry.JobKind)
		}
		for _, reason := range entry.FailureReasons {
			if _, reasonErr := ParseJobFailureReason(jobKind, reason); reasonErr != nil {
				t.Fatal(reasonErr)
			}
		}
	}
	if _, err := ParseJobKind("projection"); err == nil {
		t.Fatal("unknown job kind accepted")
	}
	if _, err := ParseJobFailureReason(JobKindMeshOptimization, "failed"); err == nil {
		t.Fatal("unknown job failure reason accepted")
	}
	if _, err := ParseJobFailureReason("projection", "failed"); err == nil {
		t.Fatal("unknown job kind accepted for failure reason")
	}
	if got := JobKindMeshOptimization.String(); got != "mesh_optimization" {
		t.Fatalf("String() = %q", got)
	}
}
